File: LuaProj/modules/ShellSort.lua
ShellSort = {}

local data = {}

local function InitData()
    print("Init Data")
    for i = 0, 20 do 
        data[i] = math.random( 0, 100)
        print("data[",i,"] = ", data[i])
    end
end

function div2(x)
    x = x / 2
    return x
end

function Check1(gap)
    return 1
end

function Check(x, y, data)
    return x >= y and data[x - y] < 0
end

function minusGap(j, gap)
    return j - gap
end

function ShellSort.Sort()
    InitData()
    print("Start Shell Sort")    
    local gap = math.modf(#data / 2)
    while(gap > 0)
    do
        for i = gap, #data do
            local temp = data[i]
            j = i
            while(j >= gap and temp <= data[j- gap])
            do            
              data[j] = data[j - gap]
              j = j - gap
            end
            data[j] = temp
        end
        gap = math.modf(gap /  2)
    end    
    printArry(data)
    print("End Shell Sort")
end

return ShellSort

File: CodeTest/shellsort.cpp
#include "shellsort.h"
/*
希尔排序，插入排序的变种
通过比较相隔一定距离的元素，每趟排序的间隔，逐步减小，直到只比较相邻的元素
https://baike.baidu.com/item/%E5%B8%8C%E5%B0%94%E6%8E%92%E5%BA%8F/3229428?fr=aladdin
*/
void shellsort::sort(vector<int>& data)
{
	int gap, i, j;
	int temp;
	for (gap = data.size() >> 1; gap > 0; gap >>= 1)
	{
		for (i = gap; i < data.size(); i++)
		{
			temp = data[i];
			//
			for (j = i - gap; j >= 0 && data[j] > temp; j -= gap)
			{
				data[j + gap] = data[j];
			}

			data[j + gap] = temp;
		}
	}
}
